xargs/xargs.1
.\"
.\" Copyright (c) 2003 Gunnar Ritter
.\"
.\" SPDX-Licence-Identifier: Zlib
.\"
.\" Sccsid @(#)xargs.1	1.8 (gritter) 8/14/05
.TH XARGS 1 "8/14/05" "Heirloom Toolchest" "User Commands"
.SH NAME
xargs \- construct argument list(s) and execute command
.SH SYNOPSIS
\fBxargs\fR [\fIflags\fR] [\fIcommand\fR\ [\fIinitial-arguments\fR]]
.SH DESCRIPTION
.I Xargs
executes
.I command
using the given
.I "initial arguments"
followed by arguments read from standard input.
By default,
.I command
is invoked as many times as necessary
using the maximum number of arguments possible
for each invocation.
If no
.I command is given,
.I echo
is used.
.PP
Data read from standard input is split into a list of arguments,
using blanks and newline characters as delimiters.
Special characters are: backslash
.RB ( \e ),
which causes any following character to be interpreted literally,
and single (\') and double (") quotes,
which cause following characters except newline
to be taken literally until the quote character appears again.
.PP
The following option flags are accepted:
.TP 12
\fB\-e\fR[\fIeofstr\fR]
Sets the logical end-of-file delimiter to
.IR eofstr ;
data is read from standard input
until an argument equal to
.I eofstr
is encountered.
If
.I eofstr
is omitted, no logical end-of-file delimiter is used
and data is read from standard input until the physical
end-of-file is reached.
The default logical end-of-file delimiter is
a single underscore character
.RB ( _ ).
.TP 12
\fB\-i\fR[\fIreplstr\fR]
Changes the operation of
.I xargs
such that each line read from standard input
replaces occurrences of
.I replstr
within
the
.IR initial-arguments .
If
.I replstr
is omitted,
.B {}
is default.
.\"No more than five
Any member of
.I initial-arguments
can each contain any number of occurrences of
.IR replstr .
The
.I command
is executed once for each line
until physical end-of-file is reached
or a line consisting exactly of the
logical end-of-file delimiter is read.
Empty lines and leading blank characters are ignored,
other blank characters are kept regardless of quotes.
.TP 12
\fB\-l\fR[\fInumber\fR]
Invokes command for each
.I number
of lines;
1 is default if no
.I number
is specified.
Empty lines are ignored;
if a line ends with an unquoted blank character,
it is not counted.
.TP 12
\fB\-n\fI\ number\fR
Executes command for each
.I number
of arguments read from standard input,
or with fewer arguments if the maximum argument length is exceeded
or fewer than
.I number
arguments remain.
.TP 12
.B \-p
Each resulting command is written to standard error
along with its arguments
before it is executed
and the user is asked whether it should be executed.
If the response is affirmative,
the command is executed;
otherwise, the execution of this command instance is skipped.
.TP 12
\fB\-s\fI\ size\fR
Sets the maximum length of generated command lines to
.IR size .
.TP 12
.B \-t
Trace mode;
each resulting command is written to standard error
along with its arguments
before it is executed.
.TP 12
.B \-x
If execution cannot be done within the limits specified by the
.BR \-n ,
.BR \-L ,
or
.BR \-s
flags,
.I xargs
will terminate.
.PP
The following options have been introduced by X/Open:
.TP 12
\fB\-E\fI\ eofstr\fR
Same as
\fB\-e\fIeofstr\fR.
.TP 12
\fB\-L\fI\ number\fR
Same as
\fB\-l\fInumber\fR.
.TP 12
\fB\-I\fI\ replstr\fR
Same as
\fB\-i\fIreplstr\fR.
.PP
When any instance of
.I command
exits with a status of \-1
or is terminated by a signal,
.I xargs
terminates.
.SH "ENVIRONMENT VARIABLES"
.TP
.BR LANG ", " LC_ALL
See
.IR locale (7).
.TP
.B LC_CTYPE
Determines the mapping of bytes to characters
and the set of blank characters.
.TP
.B PATH
Used to locate
.IR command .
.SH FILES
/dev/tty
Used with the
.B \-p
option.
.SH "SEE ALSO"
find(1),
sh(1),
locale(7)
.SH NOTES
When
.I xargs
is used to read data
generated by the
.I find
or
.I ls
commands and any of the characters special to
.I xargs
appear in file names,
.I command
will be invoked with a wrong set of file names.
To avert this,
.I find
is given a special pathname consisting of two slashes
that marks the beginning of a file name,
and each character is preceded with a backslash:
.RS
.nf
.sp
find .//. \-print | sed \'s/./\e\e&/g\' | awk \'{
        if (NR > 1) {
                printf("%s", line)
                if ($0 !~ /\e\e\e/\e\e\e//)
                        printf("\e\e")
                printf("\en")
        }
        line = $0
}
END {
        print(line)
}\' | xargs \fIcommand\fR
.fi
.sp
.RE
Note that the
.RI ` "\-exec command {} +" '
operand of
.I find
provides a method to achieve similar results
as with
.I xargs
for simple cases.


CHANGES
Release ...
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants. 

Release 230503 
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin. 

Release 230120
* Notes for the library directories if linking against the GNU C library.
* rm: Consistency within standards: the SUS program shouldn't have the POSIX 1003.1(2008) fix, just the main program.
* Made $(ROOT) not hardcoded anymore.
* Cleaned up more licence headers (if my estimations are correct, we've got rid of 3.255 lines!).
* Small update on README.md --- although it's now less readable on plain-text readers such as more or pg.

Release 230107
* rm: Added support for the POSIX 1003.1(2008).
* chroot: Added manual page.
* awk: Added stdint.h to awk.g.y.
* Now it has a bigger range of compatibility with GNU C library versions.

Release 221025
* Made notes clearer about building Heirloom on modern versions of the GNU C library.
* tar: Support for zstd and xz.

Release 220821
* [NEW] chroot: New implementation of chroot, based on Solaris 2.5 payloads.
* true, false: Rewritten in C.
* yes: Rewritten and adapted to ANSI C.
* "libcommon/vpfmt.c": Adapted to ANSI C.
* Now many of the utilities can be built in the musl C library and GNU C library (dynamically).

Release 070715
* cpio: A "-P" option has been introduced that allows to specify attributes
  like the file owner when creating an archive. With it, unprivileged users
  can set arbitrary attributes for files within an archive.
* diff: Made it compile on CentOS 5.
* Static linking against glibc is now possible since "optopt" is no longer
  a common symbol, and there is no conflict anymore with "getopt.o".
* cpio, tar: A compilation problem on some Linux systems has been fixed.
* ul: A compilation problem for NetBSD and other systems that use termcap
  instead of curses has been fixed (patch by Aleksey Cheusov).
* bfs: An OpenSolaris-derived manual page is now included.

Release 070227
* The directory for development utilities, by default "/usr/ccs/bin", is
  now configurable in "build/mk.config" . It is expected that the utilities
  from the "Heirloom Development Tools" package are installed in it.
* getconf: "/usr/ccs/bin" is now included in the default path as printed
  with the "PATH" argument.
* what: Has been replaced by an OpenSolaris-derived implementation.
* tsort: Is now installed in "/usr/ccs/bin".
* OpenSolaris-derived manual pages are now included for bdiff, newform,
  and fspec(5).
* psrinfo: Fixed the computation of the "on-line since" time for Linux 2.6.

xargs/xargs.c
/*
 * xargs - construct argument list(s) and execute command
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, March 2003.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * SPDX-Licence-Identifier: Zlib
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
static const char sccsid[] USED = "@(#)xargs.sl	1.15 (gritter) 6/21/05";

#include	<sys/types.h>
#include	<sys/stat.h>
#include	<sys/wait.h>
#include	<fcntl.h>
#include	<unistd.h>
#include	<stdio.h>
#include	<string.h>
#include	<stdlib.h>
#include	<libgen.h>
#include	<locale.h>
#include	<wchar.h>
#include	<wctype.h>
#include	<ctype.h>
#include	<errno.h>

#include	<iblok.h>
#include	<blank.h>
#include	<mbtowi.h>

#if defined (__GLIBC__) && defined(_IO_putc_unlocked)
#undef	putc
#define	putc(c, f)	_IO_putc_unlocked(c, f)
#endif

#define	peek(wc, s, n)	(mb_cur_max > 1 && *(s) & 0200 ? \
			((n) = mbtowi(&(wc), (s), mb_cur_max), \
		 	(n) = ((n) > 0 ? (n) : (n) < 0 ? (wc=WEOF, 1) : 1)) :\
		((wc) = *(s) & 0377, (n) = 1))

/*
 * Wording of SUSv2 and SUSv3 seems to require some static limits. This
 * implementation can go beyond, but if you want to have a bone-headed
 * standards-compliant utility, #define WEIRD_LIMITS.
 */
#undef	WEIRD_LIMITS

static const char	*progname;
static const char	*eflag = "_";
static const char	*iflag;
static size_t		iflen;
static long		lflag;
static long		nflag;
static int		pflag;
static const char	*sflag;
static int		tflag;
static int		xflag;
static int		errcnt;
static int		mb_cur_max;

static long		a_cnt;		/* count of arguments */
static long		a_agg;		/* aggregate commands */
static long		a_cur;		/* current position in aggregate */
static long		a_asz;		/* aggregate command length */
static long		a_csz;		/* aggregate current length */
static long		a_maxsize;	/* aggregate maximum length */
static const char	**a_vec;	/* arguments */
static char		*a_spc;		/* aggregate space */
static long		a_maxarg;	/* maximum arguments in e_vec */

static long		*replpos;	/* position of arguments with replstr */
static long		replsiz;	/* number of members in replpos */
static int		replcnt;	/* count of arguments with replstr */

static void *
srealloc(void *op, size_t size)
{
	void	*np;

	if ((np = realloc(op, size)) == NULL) {
		write(2, "no memory\n", 10);
		_exit(077);
	}
	return np;
}

static void *
smalloc(size_t size)
{
	return srealloc(NULL, size);
}

static const char *
contains(const char *haystack, const char *needle)
{
	size_t	ln = strlen(needle);
	wint_t	wc;
	int	n;

	while (*haystack) {
		if (*haystack == *needle && strncmp(haystack, needle, ln) == 0)
			return haystack;
		peek(wc, haystack, n);
		haystack += n;
	}
	return NULL;
}

static long
envsz(void)
{
	extern char	**environ;
	long	val = 0;
	int	i;

	for (i = 0; environ[i]; i++)
		val += strlen(environ[i]) + 1;
	return val;
}

static const char *
validate(const char *s, int optc)
{
	const char	*sp = s;
	wint_t	wc;
	int	n;

	if (mb_cur_max > 1) {
		while (*sp) {
			peek(wc, sp, n);
			if (wc == WEOF) {
				fprintf(stderr,
			"%s: illegal byte sequence in argument to -%c\n",
					progname, optc);
				exit(1);
			}
			sp += n;
		}
	}
	return s;
}

static void
overflow(void)
{
	if (iflag)
		fprintf(stderr,
			"%s: max arg size with insertion via {}'s exceeded\n",
			progname);
	else
		fprintf(stderr, "%s: arg list too long\n", progname);
	exit(1);
}

static void
unknown(const char *s)
{
	fprintf(stderr, "%s: unknown option: -%s\n", progname, s);
	exit(1);
}

static void
missing(int c)
{
	fprintf(stderr, "%s: missing argument: -%c\n", progname, c);
	exit(1);
}

static void
mustbepos(int c, const char *s)
{
	const char	*type = 0;

	switch (c) {
	case 'l':
	case 'L':
		type = "#lines";
		break;
	case 'n':
		type = "#args";
		break;
	}
	fprintf(stderr, "%s: %s must be positive int: -%c%s\n",
			progname, type, c, s);
	exit(1);
}

static long
number(int c, const char *s)
{
	char	*x;
	long	val;

	val = strtol(s, &x, 10);
	if (val <= 0)
		mustbepos(c, s);
	return val;
}

static int
flags(int ac, char **av)
{
	int	i;

	for (i = 1; i < ac && av[i][0] == '-'; i++) {
	nxt:	switch (av[i][1]) {
		case 'e':
			eflag = validate(&av[i][2], 'e');
			continue;
		case 'E':
			if (av[i][2])
				eflag = validate(&av[i][2], 'E');
			else if (++i < ac)
				eflag = validate(av[i], 'E');
			else
				missing('E');
			continue;
		case 'i':
			if (av[i][2])
				iflag = validate(&av[i][2], 'i');
			else
				iflag = "{}";
			continue;
		case 'I':
			if (av[i][2])
				iflag = validate(&av[i][2], 'i');
			else if (++i < ac)
				iflag = validate(av[i], 'i');
			else
				missing('I');
			continue;
		case 'l':
			if (av[i][2])
				lflag = number('l', &av[i][2]);
			else
				lflag = 1;
			xflag = 1;
			iflag = NULL;
			continue;
		case 'L':
			if (av[i][2])
				lflag = number('L', &av[i][2]);
			else if (++i < ac)
				lflag = number('L', av[i]);
			else
				mustbepos('L', "");
			xflag = 1;
			iflag = NULL;
			nflag = 0;
			continue;
		case 'n':
			if (av[i][2])
				nflag = number('n', &av[i][2]);
			else if (++i < ac)
				nflag = number('n', av[i]);
			else
				mustbepos('n', "");
			lflag = 0;
			continue;
		case 'p':
			pflag = open("/dev/tty", O_RDONLY);
			fcntl(pflag, F_SETFD, FD_CLOEXEC);
			tflag = 1;
			break;
		case 's':
			if (av[i][2])
				sflag = &av[i][2];
			else if (++i < ac)
				sflag = av[i];
			else
				sflag = "0";
			continue;
		case 't':
			tflag = 1;
			break;
		case 'x':
			xflag = 1;
			break;
		case '-':
			return ++i;
		default:
			unknown(&av[i][1]);
		}
		if (av[i][2]) {
			(av[i])++;
			goto nxt;
		}
	}
	return i;
}

static void
run(const char **args)
{
	pid_t	pid;
	int	status;

	if (tflag) {
		int	i;
		const char	*cp;

		for (i = 0; args[i]; i++) {
			if (i > 0)
				putc(' ', stderr);
			for (cp = args[i]; *cp; cp++)
				putc(*cp & 0377, stderr);
		}
		if (pflag) {
			char	ch, ch2;

			for (cp = " ?..."; *cp; cp++)
				putc(*cp, stderr);
			fflush(stderr);
			switch (read(pflag, &ch, 1)) {
			case -1:
				fprintf(stderr,
					"%s: can't read from tty for -p\n",
					progname);
				exit(1);
				/*NOTREACHED*/
			case 0:
				exit(errcnt);
			}
			if (ch != '\n')
				while (read(pflag, &ch2, 1) == 1 &&
						ch2 != '\n');
			if (ch != 'y')
				return;
		} else
			putc('\n', stderr);
	}
	switch (pid = fork()) {
	case -1:
		fprintf(stderr, "%s: cannot fork(): %s\n", progname,
				strerror(errno));
		exit(126);
		/*NOTREACHED*/
	case 0:
		execvp(args[0], (char **)args);
		_exit(errno == ENOENT ? 127 : 126);
	default:
		while (wait(&status) != pid);
		if (status && ((WIFEXITED(status)&&WEXITSTATUS(status)==255) ||
				WIFSIGNALED(status))) {
			fprintf(stderr, "%s: %s not executed or returned -1\n",
					progname, args[0]);
			exit(125);
		}
		if (status)
			errcnt |= WIFEXITED(status) ? WEXITSTATUS(status) : 1;
	}
}

static void
addcmd(const char *s)
{
	if (a_cnt >= a_maxarg - 4096) {
		a_maxarg += 8192;
		a_vec = srealloc(a_vec, a_maxarg * sizeof *a_vec);
	}
	if (iflag && contains(s, iflag)) {
		if (replcnt >= replsiz) {
#ifdef	WEIRD_LIMITS
			if (replcnt >= 5) {
				fprintf(stderr, "%s: too many args with %s\n",
					progname, iflag);
				exit(1);
#endif	/* WEIRD_LIMITS */
			replpos = srealloc(replpos,
					(replsiz += 5) * sizeof *replpos);
		}
		replpos[replcnt++] = a_cnt;
	}
	a_vec[a_cnt++] = s;
	a_asz += strlen(s) + 1;
}

static void
endcmd(void)
{
	a_agg = a_cnt;
	a_maxsize = sysconf(_SC_ARG_MAX) - envsz() - 2048 - a_asz;
	if (nflag || sflag) {
		long	newsize = sflag ? atol(sflag) :
#ifdef	WEIRD_LIMITS
			LINE_MAX;
#else	/* !WEIRD_LIMITS */
			a_maxsize;
#endif	/* !WEIRD_LIMITS */
		if (sflag && (newsize <= a_asz || newsize > a_maxsize))
			fprintf(stderr,
				"%s: 0 < max-cmd-line-size <= %ld: -s%s\n",
				progname, a_maxsize, sflag);
		if (newsize < a_maxsize)
			a_maxsize = newsize;
	}
	if (sflag)
		a_spc = smalloc(a_maxsize);
	else {
		while ((a_spc = malloc(a_maxsize)) == NULL) {
			a_maxsize /= 2;
			if (a_maxsize < LINE_MAX) {
				a_spc = smalloc(a_maxsize);
				break;
			}
		}
	}
	a_cur = a_agg;
	if (nflag && nflag < a_maxarg + a_cnt) {
		nflag += a_cnt;
		a_maxarg = nflag;
	}
}

static void
addarg(const char *s, int always)
{
	size_t	sz;
	int	toolong = 0;
	static int	didexec;

	do {
		if (s) {
			sz = strlen(s) + 1;
			if (a_csz == 0 && sz > a_maxsize) {
				fprintf(stderr,
		"%s: a single arg was greater than the max arglist size\n",
					progname);
				exit(1);
			}
			if (a_csz + sz <= a_maxsize && a_cur < a_maxarg) {
				strcpy((char *)(a_vec[a_cur++]=&a_spc[a_csz]),
						s);
				a_csz += sz;
				if (always)
					goto doit;
				return;
			} else {
				if (a_agg == a_cur)
					a_vec[a_cur++] = s;
				else
					toolong = 1;
			doit:	if (nflag && xflag && a_cnt > nflag)
					overflow();
				if (lflag && xflag && !always)
					overflow();
				always = 0;
			}
		} else {
			if ((didexec || lflag) && a_agg == a_cur)
				return;
		}
		didexec = 1;
		a_vec[a_cur] = NULL;
		a_cur = a_agg;
		a_csz = 0;
		run(a_vec);
	} while (toolong != 0);
}

static char *
replace(const char *arg, const char *s)
{
	wint_t	wc;
	int	n;
	size_t	sz = strlen(s);
	char	*bot = &a_spc[a_csz];

	while (*arg != '\0') {
		if (*arg == *iflag && strncmp(arg, iflag, iflen) == 0) {
			if (a_csz + sz > a_maxsize)
				overflow();
			strcpy(&a_spc[a_csz], s);
			a_csz += sz;
			arg += iflen;
		} else {
			peek(wc, arg, n);
			if (a_csz + n > a_maxsize)
				overflow();
			while (n--)
				a_spc[a_csz++] = *arg++;
		}
	}
	if (a_csz + 1 > a_maxsize
#ifdef	WEIRD_LIMITS
			|| &a_spc[a_csz] - bot > 255
#endif	/* WEIRD_LIMITS */
			)
		overflow();
	a_spc[a_csz++] = '\0';
	return bot;
}

static void
insert(const char *s)
{
	const char	**args;
	int	i, j;

	args = smalloc((a_cnt + 1) * sizeof *args);
	a_csz = 0;
	for (i = 0, j = 0; i < a_cnt; i++) {
		if (j < replcnt && replpos[j] == i) {
			j++;
			args[i] = replace(a_vec[i], s);
		} else {
			a_csz += strlen(a_vec[i]) + 1;
			if (a_csz > a_maxsize)
				overflow();
			args[i] = a_vec[i];
		}
	}
	args[i] = NULL;
	run(args);
	free(args);
}

#define	nextc()		(cp = (mb_cur_max > 1 ? ib_getw(ip, &c, &n) : \
		(c = ib_get(ip)) == (wint_t)EOF ? NULL : (b = c, n = 1, &b)))
			
#define	blankc(c)	(mb_cur_max > 1 ? iswblank(c):isblank(c))

static char *
nextarg(struct iblok *ip, long *linecnt)
{
	static char	*buf;
	static size_t	size;
	static int	eof;
	wint_t	c, quote = WEOF, lastc = WEOF;
	char	*cp;
	char	b;
	int	content = 0, i = 0, n;

	if (eof)
		return NULL;
	if (buf == NULL)
		buf = smalloc(size = 128);
	for (;;) {
		nextc();
		if (cp) {
			if (c != WEOF && c == quote) {
				quote = WEOF;
				continue;
			}
			lastc = c;
			if (quote == WEOF && c == '\\') {
				content = 1;
				nextc();
			} else {
				if (quote == WEOF && (c == '"' || c == '\'')) {
					quote = c;
					content = 1;
					continue;
				}
				if (c == '\n' || quote == WEOF &&
						blankc(c) &&
						(iflag == NULL ||
						 content == 0)) {
					if (content) {
						if (c == '\n' && !blankc(lastc))
							(*linecnt)++;
						break;
					} else
						continue;
				}
			}
		}
		if (cp == NULL)
			break;
		content = 1;
		while (n--) {
			buf[i] = *cp++;
			if (++i >= size)
				buf = srealloc(buf, size += 128);
		}
	}
	buf[i] = '\0';
	if (quote != WEOF) {
		fprintf(stderr, "%s: missing quote?: %s\n", progname, buf);
		exit(1);
	}
	if (cp == NULL)
		eof = 1;
	return content ? buf : NULL;
}

static void
process(void)
{
	struct iblok	*ip;
	char	*arg;
	long	linecnt = 0;

	ip = ib_alloc(0, 0);
	if (iflag) {
		iflen = strlen(iflag);
		while ((arg = nextarg(ip, &linecnt)) != NULL) {
			if (eflag && *eflag && strcmp(arg, eflag) == 0)
				break;
			insert(arg);
		}
	} else {
		do {
			arg = nextarg(ip, &linecnt);
			if (arg && eflag && *eflag && strcmp(arg, eflag) == 0)
				arg = NULL;
			addarg(arg, lflag && lflag == linecnt ?
					linecnt = 0, 1 : 0);
		} while (arg != NULL);
	}
}

int
main(int argc, char **argv)
{
	static char	errbuf[256];
	int	i;

	setvbuf(stderr, errbuf, _IOLBF, sizeof errbuf);
	progname = basename(argv[0]);
	setlocale(LC_CTYPE, "");
	mb_cur_max = MB_CUR_MAX;
	i = flags(argc, argv);
	argc -= i, argv += i;
	if (argc > 0)
		for (i = 0; i < argc; i++)
			addcmd(argv[i]);
	else
		addcmd("echo");
	endcmd();
	process();
	return errcnt;
}
